Begin instruction for differential expression notebook (#592)
AlexsLemonade/training-modules: scRNA-seq-advanced/04-differential_expression.Rmd

    geom_bar(color = "black", size = 0.2) +
    labs(
      x = "Sample",
      y = "Number of cells", 
      fill = "Cell type"
    ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))+
  facet_wrap(~diagnosis_group, scales = "free_x")
```

*Need to explain the difference between tumor/tumor_myoblast, etc*
These plots show that ARMS and ERMS share a lot of the same cell types, but perhaps there are differences in the cells in each subdiagnoses.
We also see that 6 of the libraries show a mixture of tumor cells belonging to the mesoderm, myoblast, and myocyte classes.
3 libraries contain cells that are only classified as tumor or non-tumor, and tumor cells are not further classified and the remaining library appears to not have been assigned any cell types, as they are all labeled with `NA`.
We will not be using these 3 libraries.

```{r}
# define libraries to keep
library_ids <- c(
  "SCPCL000479",
  "SCPCL000480",
  "SCPCL000481",
  "SCPCL000484",
  "SCPCL000488",
  "SCPCL000491"
)

# subset sce to only contain samples with library IDs of interest 
samples_to_keep <- integrated_sce$sample %in% library_ids
rms_sce <- integrated_sce[,samples_to_keep]

> ```

Reviewer: It looks like this notebook got killed in CI for memory usage. We probably want to get rid of the integrated object as soon as we subset it here. If that doesn't work, we may need to think about other solutions.

  dplyr::left_join(sce_rowdata_df, by = "ensembl_id")
```

```{r}
# save our results 
readr::write_tsv(deseq_results, deseq_output_file)
```

### Exploring the identified differentially expressed genes 

Let's also do some initial analysis to look at what types of genes are being identified and see if what is being identified as actually up/down regulated appear to correlate back to the single cell data. 

The first plot we'll make is a volcano plot and we can label the significant genes with their gene symbols. 

```{r}
EnhancedVolcano::EnhancedVolcano(deseq_results,
                x = 'log2FoldChange', # fold change statistic to plot
                y = 'pvalue', # significance values
                lab = deseq_results$gene_symbol, # labels for points
                pCutoff = 1e-05, # The p value cutoff we will use (default)
                FCcutoff = 1, # The fold change cutoff (default)
                title = NULL, # no title
                subtitle = NULL, # or subtitle
                caption = NULL, # or caption
                drawConnectors = TRUE, # add some fun arrows
                labSize = 3  # smaller labels
                ) +
  # change the overall theme

>   theme_classic() +

Reviewer 2: `theme_bw()` maybe? I think that will be more consistent with other notebooks.

---
title: "Differential expression analysis for single-cell RNA-seq"
author: "Data Lab for ALSF"
date: "2022"
output:
  html_notebook:
    toc: yes
    toc_float: yes
---

## Objectives 

This notebook will demonstrate how to:

- Use pseudo-bulking to prepare `SingleCellExperiment` objects for differential expression
- Perform differential expression with the `DESeq2` package
- Compare cell type abundance across samples

---

In this notebook, we will perform differential expression analysis across a set of samples using the `DESeq2` package. 
We will continue working with samples from the [`SCPCP000005` project](https://scpca.alexslemonade.org/projects/SCPCP000005) we worked with in the integration notebook.
This particular dataset contains 10 different samples that have been integrated using `fastMNN`, following the same procedure we outlined in `03-dataset_integration.Rmd`.

*Need to spend some time here introducing DE analysis and the packages we will be using* 

## Set Up

> ```{r}

Author: Noted that this will be added later in https://github.com/AlexsLemonade/training-modules/issues/573#issuecomment-1353398710.

We can also look at the distribution of cell types to see if these sub types have different or shared cell types.

*Talk about looking across cell types and comparing cell types* 

```{r}
# facet the UMAP by sample to show the distribution of cell types across each subdiagnosis
scater::plotReducedDim(integrated_sce,
                       dimred = "fastmnn_UMAP",
                       colour_by = "celltype_broad",
                       point_size= 0.5, 
                       point_alpha = 0.4,
                       other_fields = "subdiagnosis") +
  facet_wrap(~ subdiagnosis)

# we can also look at the sample level
scater::plotReducedDim(integrated_sce,
                       dimred = "fastmnn_UMAP",
                       colour_by = "celltype_broad",
                       point_size= 0.5, 
                       point_alpha = 0.4,
                       other_fields = "sample") +
  facet_wrap(~ sample)
```

We can use these plots to help explore all of the samples and visualize the distribution of cell types across our dataset.
*Need to explain the difference between tumor/tumor_myoblast, etc*

> These plots show that ARMS and ERMS share a lot of the same cell types, but perhaps there are differences in the cells in each sub type.

Author: Just noting in general that I plan on re-wording/ updating text in subsequent PRs for each section as noted in https://github.com/AlexsLemonade/training-modules/issues/573#issuecomment-1353398710. So not ignoring these comments, just saving for later!

samples_to_keep <- integrated_sce$sample %in% library_ids
rms_sce <- integrated_sce[,samples_to_keep]
```

```{r}
# we are left with 3 from each sub type 
sample_metadata %>%
  dplyr::filter(library_id %in% library_ids) %>%
  dplyr::pull(subdiagnosis) %>%
  table()
```

```{r}
# look at distribution of cell types across all libraries 
as.data.frame(colData(rms_sce)) %>%
  dplyr::count(celltype_broad)
```

The `Tumor_Myoblast` cell type is the most prominent cell type, and based on the UMAPs is found in all 6 of our samples.
We will start by comparing the gene expression of the myoblasts across ARMS and ERMS samples.

## Differential Expression Analysis

### Pseudo-bulking 

Before we can compare the gene expression profiles of myoblasts in ARMS vs. ERMS patients, we will need to "pseudo-bulk" the gene counts. 
Pseudo-bulking creates a new counts matrix that contains the total counts across all cells of a given label (e.g. cell type) for each sample. 
This allows us to use single-cell resolution to define the labels and sum gene counts across groups of cells containing the same label, and avoids counting each cell as its own replicate. 

> Pseudo-bulking is implemented prior to differential expression analysis for the following reasons: 

Reviewer 2: ```suggestion
Pseudo-bulking is implemented prior to differential expression analysis on single-cell data for the following reasons: 
```

> - Produces larger counts, which are more amenable to standard normalization and differential expression methods used by bulk RNA-sequencing. 

Reviewer 2: up to you - 
```suggestion
- Produces larger counts, which allows us to use standard normalization and differential expression methods used by bulk RNA-sequencing. 
```

-  Collapses gene expression counts by sample, so that samples, rather than cells, represent replicates.

> - Masks variance within a sample to highlight variance across samples.

Reviewer 2: ```suggestion
- Masks variance within a sample to emphasize variance across samples.
```

> We are staring with the simple comparison of looking at one cell type and comparing across two groups of samples, samples belonging to either the ERMS and ARMS subtypes. 

Reviewer 2: ```suggestion
We are starting with the simple comparison of looking at one cell type and comparing across two groups of samples: the ERMS and ARMS subdiagnoses. 
```

> When we do this, we will first want to pseudo-bulk our dataset to group our cells by cell type and by sample.

Reviewer 2: ```suggestion
Our first step will be to pseudo-bulk our dataset to group our cells by cell type and by sample.
```

> This will create a new `SingleCellExperiment` object that contains the pseudo-bulked counts to be used as input for differential expression analysis. 
> Then we can subset the pseudo-bulked object to just contain our cell type of interest before proceeding. 

Reviewer 2: ```suggestion
This will create a new `SingleCellExperiment` (SCE) object that contains the pseudo-bulked counts for all genes. 
We can subset this SCE to just contain our cell type of interest (tumor myoblasts) for input to differential expression analysis. 
```

> Before we apply pseudo-bulking to our dataset, lets look at a simple example of how pseudo-bulking works. 

Reviewer 2: ```suggestion
Before we apply pseudo-bulking to our dataset, let's look at a simple example of how pseudo-bulking works. 
```

```{r}
# create an example counts matrix
counts_mtx <- matrix(1:12, 
                     ncol = 4,
                     dimnames = list(c("geneA", "geneB", "geneC"),
                                     c("A-cell1", "A-cell2", "B-cell1", "B-cell2")))
counts_mtx
```

```{r}
# create new matrix, summing the counts for each gene across all cells in the specified group
pseudobulked_counts <- cbind(

>   rowSums(counts_mtx[,1:2]), # sum all counts for cells labeled with group A

Reviewer 2: a bit more explicit - 
```suggestion
  rowSums(counts_mtx[,c("A-cell1", "A-cell2")]), # sum all counts for cells labeled with group A
```

>   rowSums(counts_mtx[,3:4]) # sum all counts for cells labeled with group B

Reviewer 2: ```suggestion
  rowSums(counts_mtx[,c("B-cell1", "B-cell2")]) # sum all counts for cells labeled with group B
```

)
pseudobulked_counts 
```
Now the actual pseudo-bulking for our dataset! 

> We can pseudo-bulk using any groups that we are interested in.

Reviewer 2: ```suggestion
We can pseudo-bulk using any grouping that we are interested in.
```

> For right now, we would are interested in looking at gene expression across cell types, so we want to group the counts matrix by both cell type and sample. 

Reviewer 2: ```suggestion
For right now, we would are interested in looking at gene expression across cell types, so we want to group the pseudo-bulked counts matrix by both cell type and sample. 
```

```{r}

> # first let's subset our coldata to only have the columns we care about in pseudobulking 

Reviewer 2: ```suggestion
# first let's subset our coldata to only have the columns we care about in pseudo-bulking 
```

pseudobulked_coldata <- colData(rms_sce)[, c("celltype_broad", "sample")]

> # this creates a new SCE object that contains the pseudobulked counts across the provided groups 

Reviewer 2: ```suggestion
# this creates a new SCE object that contains the pseudo-bulked counts across the provided groups 
```

Reviewer 2: Small note, we'll want to really emphasize that the _counts_ get summed, not the logcounts (by default).

pseudobulked_sce <- scater::aggregateAcrossCells(rms_sce, 
                                                  id = pseudobulked_coldata)

# column names aren't automatically added, so let's add them in 
colnames(pseudobulked_sce) <- paste(pseudobulked_sce$celltype_broad, 
                                    pseudobulked_sce$sample, sep = "-")

pseudobulked_sce
```

How does the new pseudo-bulked `SingleCellExperiment` look different? 

```{r}
# only 37 columns in the counts assay 
counts(pseudobulked_sce)[1:10, 1:10]
```

Let's take a look at what the `colData` now looks like in the pseudo-bulked SCE object. 

```{r}
# note the new addition of the column with number of cells per group 
colData(pseudobulked_sce)
```

### Filtering the pseudo-bulked `SingleCellExperiment` object

*Add a bit more justification on filtering and why we need it*

```{r}
# filter to only the HVG list

> hv_genes <- metadata(integrated_sce)$combined_hvg

Author: Yea I wasn't entirely sure about this one... Looking at the OSCA chapter they show the DE analysis example on an object that has been merged and corrected and only contains the HVGs. But the way they choose HVGs is slightly different, instead of using `getTopHVGs` they retain anything with a positive biological component of the variance. So they retain a lot more than 2000 genes. 
Here's the link to how they set up the data that they used for reference: http://bioconductor.org/books/3.16/OSCA.multisample/chimeric-mouse-embryo-10x-genomics.html#chimeric-mouse-embryo-10x-genomics

I think filtering to only the 2000 genes may be too stringent, but I think we definitely want to do some sort of filtering.

> hv_genes <- metadata(integrated_sce)$combined_hvg

Reviewer: I might try it without filtering... I think the main reason for filtering is to speed calculations, but I don't expect that to be too bad.

filtered_pseudobulked_sce <- pseudobulked_sce[hv_genes, ]
```

We want to remove any groups in the pseudo-bulked dataset that may have a low number of cells. 

```{r}
# filter by number of cells 

> filtered_pseudobulked_sce <- filtered_pseudobulked_sce[, pseudobulked_sce$ncells >= 10]

Reviewer 2: Might want to talk about "why 10" here

```

We also want to filter out any genes that may be lowly expressed in our pseudo-bulked dataset.

*Talk about how this specific function uses the sample groups provided to determine the minimum sample size that genes are expressed in.*

```{r}

> # filter out genes that might be lowly expressed using filterbyExpr in `edgeR`

Reviewer 2: might want to note here that we're not actually using edgeR for the differential expression but it has useful utility functions

# filter by genes
filtered_pseudobulked_sce <- filtered_pseudobulked_sce[genes_filter,]

```

```{r}
dim(pseudobulked_sce)
dim(filtered_pseudobulked_sce)
```

### Perform differential expression with DESeq2

*Intro about DESeq2*

Before we identify differentially expressed genes, we will want to subset the pseudo-bulked `SingleCellExperiment` object to contain only the cell type that we are interested in comparing across the two sub types.

```{r}
tumor_myoblast_sce <- filtered_pseudobulked_sce[, filtered_pseudobulked_sce$celltype_broad == "Tumor_Myoblast"]
```

This leaves us with 6 columns in our dataset that should all have `celltype_broad = Tumor_Myoblast`. 

```{r}
table(tumor_myoblast_sce$celltype_broad)
```

Now we are ready to set up our `DESeq2` object. 

```{r}
# deseq2 requires the coldata object as a separate input
coldata_df <- as.data.frame(colData(tumor_myoblast_sce))
  
# set up the deseq object 
deseq_object <- DESeq2::DESeqDataSetFromMatrix(counts(tumor_myoblast_sce),
                                               colData = coldata_df,
                                               design = ~ diagnosis_groups)
```

> Use the median of ratios method for count normalization followed by regularized log transformation.

Reviewer 2: Again we probably want to reiterate here that we have to normalize "again" because we summed up the raw counts and now have "new" raw counts to normalize. The previous single-cell-resolution logcounts are no longer applicable in this analysis.

```{r}
# estimate size factors first 
deseq_object <- estimateSizeFactors(deseq_object)

# normalize and log transform to use for visualization
normalized_object <- DESeq2::rlog(deseq_object, blind = TRUE, fitType = 'local')
normalized_object
```

Evaluate QC here using PCA. 

> Here we can label by the diagnosis groups to show that the highest amount of variance is due to different diagnosis types. 

Reviewer 2: ```suggestion
Here we can label by the diagnosis groups to show that the highest amount of variance is due to different subdiagnoses. 
```

We can correct this by applying a "shrinkage" procedure, which will adjust large values with small counts downward, while preserving values with larger counts, which are likely to be more accurate.

To do this, we will use the `lfcShrink()` function, but first we need to know the name and/or position of the "coefficient" that was calculated by `DESeq()`, which we can do with the `resultsNames()` function.

```{r}
# identify position of coefficient
DESeq2::resultsNames(deseq_object)
```

```{r}
# appyly logFC shrinkage
shrink_results <- DESeq2::lfcShrink(deseq_object, res = deseq_results, coef = 2, type = "apeglm")
```

```{r}
# look at what actually happened when we applyed shrinkage (adjustment of large fold changes that are overestimated)
comparison_df <- data.frame(
  lfc_original = deseq_results$log2FoldChange,
  lfc_shrunken = shrink_results$log2FoldChange,
  logmean = log10(deseq_results$baseMean)
  )

ggplot(comparison_df, 
       aes(x = lfc_original, 
           y = lfc_shrunken, 
           color = logmean)) +
  geom_point(alpha = 0.1) +
  theme_bw() +
  scale_color_viridis_c() + 
  coord_cartesian(xlim = c(-10,10), ylim = c(-10,10))
```

> Not a ton of shrinkage, but there is some. 

Author: I am on board with cutting it. To confirm you mean not doing the shrinkage at all? I don't think we need it if it isn't really doing much.

> Not a ton of shrinkage, but there is some. 

Reviewer: I'd still do the shrinkage, as it is a good thing to do. I'd just make a brief comment about why (don't want to overestimate LFC when expression is low).

  theme_classic() +
  # move the legend to the bottom
  theme(legend.position = "bottom")
```

Let's make some UMAPs where we plot a gene that's identified to be differentially expressed in a given cell type and see what the expression of that gene is across samples. 
First we need to extract the expression for each of the genes we would like to plot from the `logcounts` matrix and combine with the `colData`.

```{r}
# add column to colData with expression of gene of interest 
# here we will actually put all of the significant genes so then we can plot any of them
sig_expressed_genes <- deseq_results %>%
  dplyr::filter(pvalue < 1e-5,
                abs(log2FoldChange) > 2) %>%
  dplyr::pull(ensembl_id)

sig_genes_counts <- logcounts(integrated_sce[sig_expressed_genes,]) %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  dplyr::mutate(cell_id = colnames(integrated_sce))
  
combined_coldata_df <- as.data.frame(colData(integrated_sce)) %>%
  dplyr::left_join(sig_genes_counts, by = "cell_id")

# add back modified coldata containing gene expression for sig genes 
colData(integrated_sce) <- DataFrame(combined_coldata_df, row.names = combined_coldata_df$cell_id)
```

> We will pick a gene and look at the expression of that gene across disease type, sample, and cell type.

Reviewer 2: ```suggestion
We will pick a gene and look at the expression of that gene across subdiagnosis, sample, and cell type.
```

Reviewer 2: In next PR(s), might want add some biological context about this gene and RMS.
